raft.go

```diff
@@ -36,38 +36,70 @@ func (a *Autopilot) NumVoters() (int, error) {
 // its all done will trigger autopilot to remove dead servers if there
 // are any. Servers added by this method will start in a non-voting
 // state and later on autopilot will promote them to voting status
-// if desired by the configured promoter.
+// if desired by the configured promoter. If too many removals would
+// be required that would cause leadership loss then an error is returned
+// instead of performing any Raft configuration changes.
 func (a *Autopilot) AddServer(s *Server) error {
 	cfg, err := a.getRaftConfiguration()
 	if err != nil {
 		a.logger.Error("failed to get raft configuration", "error", err)
 		return err
 	}
 
+	var existingVoter bool
+	var voterRemovals []raft.ServerID
+	var nonVoterRemovals []raft.ServerID
+	var numVoters int
 	for _, server := range cfg.Servers {
+		if server.Suffrage == raft.Voter {
+			numVoters++
+		}
+
 		if server.Address == s.Address && server.ID == s.ID {
 			// nothing to be done as the addr and ID both already match
 			return nil
-		} else if server.Address == s.Address || server.ID == s.ID {
-			// we know that we have a partial match  so we should remove then re-add the server.
-			if err := a.removeServer(server.ID); err != nil {
-				if server.ID == s.ID {
-					return fmt.Errorf("error removing server with duplicate ID %q: %w", server.ID, err)
-				} else {
-					return fmt.Errorf("error removing server with duplicate address %q: %w", server.Address, err)
-				}
+		} else if server.ID == s.ID {
+			// special case for address updates only. In this case we should be
+			// able to update the configuration without have to first remove the server
+			if server.Suffrage == raft.Voter || server.Suffrage == raft.Staging {
+				existingVoter = true
 			}
-
-			if server.ID == s.ID {
-				a.logger.Info("removed server with duplicate ID", "id", server.ID)
+		} else if server.Address == s.Address {
+			if server.Suffrage == raft.Voter {
+				voterRemovals = append(voterRemovals, server.ID)
 			} else {
-				a.logger.Info("removed server with duplicate address", "address", server.Address)
+				nonVoterRemovals = append(nonVoterRemovals, server.ID)
 			}
 		}
 	}
 
-	if err := a.addNonVoter(s.ID, s.Address); err != nil {
-		return err
+	requiredVoters := (numVoters / 2) + 1
```

Comment:
nitpick: I did a quick search through the rest of the autopilot code to see where else we do similar checks around maintaining quorum. I found ./state.go:202 and ./reconcile.go:252 which do similar checks. That makes me think it would be convenient to have some named functions to do these simple maths for us.

Reply:
I added one and used it here and in state.go. However the one in reconcile.go is actually different and is calculating the number of things we can remove